scripts/preconditions/preconditions.py

```diff
@@ -0,0 +1,421 @@
+        ]
+
+        req = Requirements(
+            "Required APIs on service account project",
+            parent,
+            self.REQUIRED_APIS,
+            enabled,
+        )
+
+        return req.asdict()
+
+
+class BillingAccount:
+    """
+    Test that the service account is able to link the billing account to
+    projects.
+    """
+
+    REQUIRED_PERMISSIONS = [
+        "billing.resourceAssociations.create"
+    ]
+
+    def __init__(self, billing_account):
+        self.billing_account = billing_account
+
+    def validate(self, credentials):
+        service = discovery.build(
+            'cloudbilling', 'v1',
+            credentials=credentials
+        )
+
+        body = {"permissions": self.REQUIRED_PERMISSIONS}
+        resource = "billingAccounts/" + self.billing_account
+
+        request = service.billingAccounts().testIamPermissions(
+            resource=resource,
+            body=body)
+        try:
+            response = request.execute()
+        except errors.HttpError:
+            response = {"permissions": []}
+
+        req = Requirements(
+            "Service account permissions on billing account",
+            resource,
+            self.REQUIRED_PERMISSIONS,
+            response["permissions"],
+        )
+
+        return req.asdict()
+
+
+def setup():
+    logging.basicConfig()
+    logging.getLogger().setLevel(logging.WARN)
+
+    cache_log = logging.getLogger('googleapiclient.discovery_cache')
+    cache_log.setLevel(logging.ERROR)
+
+    requests_log = logging.getLogger("requests.packages.urllib3")
+    requests_log.setLevel(logging.DEBUG)
+    requests_log.propagate = True
+
+
+def get_credentials(credentials_path):
+    credentials = service_account.Credentials.from_service_account_file(
+        credentials_path)
+
+    return credentials
+
+
+class EmptyStrAction(argparse.Action):
+    """
+    Convert empty string values parsed by argparse into None.
+    """
+    def __call__(self, parser, namespace, values, option_string=None):
+        values = None if values == '' else values
+        setattr(namespace, self.dest, values)
+
+
+def argparser():
+    parser = argparse.ArgumentParser(
+        description="""Check that Project Factory preconditions are met on the
+        provided service account, project parent, and billing account.
+        """
+    )
+
+    parser.add_argument(
+        '--verbose', required=False, default=False,
+        help='Enable verbose logging'
+    )
+    parser.add_argument(
+        '--credentials_path', required=True,
+        help='The service account credentials to check'
+    )
+    parser.add_argument(
+        '--billing_account', required=True,
+        help='The billing account to be associated with a new project'
+    )
+    parser.add_argument(
+        '--org_id', required=True, action=EmptyStrAction,
+        help='The organization ID'
+    )
+    parser.add_argument(
+        '--folder_id', required=False, action=EmptyStrAction,
```

Comment:
`folder_id` is not used to verify folder permissions.

Reply:
It's been months since I've touched this pull request so my memory is foggy, but I recall running into issues where testing for the `resourcemanager.folders.create` permission required walking the parent folders and testing for the permission. I'm happy to revisit this, or remove the check entirely - what's your thinking?

Reply:
I added the validation back in to the logic and it appears to work. I tested it against an invalid folder ID and a valid folder ID with and without the `resourcemanager.folders.get` permission.